Implement an InterfaceUserObject

arovinelli opened the pull request:
- The aim of this new class is to provide an **easy to use** user object to make computation across an interface.

- A new user obejct base class similar to the `InternalSideUserObject`. The noticeable difference is that the `InterfaceUserObject` is boundary restricted but allows to grab material properties and variable across an interface. `ComputeUserObjectsThread` and `ComputeMaterialThread` now override the `onInterface` method.  Some addition has been done to `FEProblemBase` to allow this new class to properly work.

- in general the `InterfaceUserObject` is helpful for all that situations in which a UO is needed to do computation across an Interface where the `InterfaceKernel` is used.

- This will also help in #11575 #11546 

 #closes #11647

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/199713/) on 1a0b326 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11660/site)

This comment will be updated on new commits.

arovinelli commented:
`Pthreads Test` failed because the node number is changed when --distributed-mesh is used.
However results are correct. The mesh is generated inside moose and some mesh modifiers are applied. How do I fix this?

For what concern `documentation`: I added two base classes (not registered) and two classes defined inside tests? Should I document them at all?

Commit:
Implement an interface UO based on the internal side UO features
ref #11647

arovinelli commented:
ping @permcody

permcody reviewed (commented):
A quick glance and this looks like a nice new capability - Thanks! Between training, meetings at ANL, and a milestone this week, this will have to wait until the team is back next week before we can go over it and fully evaluate it. Thanks!

arovinelli commented:
I was able to fix the segfault generated when using a stateful material property. However I'm not sure this is the best solution. To prevent the crash, I had to modify some lines in `ComputePropertyStorage` to prevent the crash. The crash was caused by looping over the stateful variable list contained into `_stateful_prop_id_to_prop_id`. even when working on the slave block from this variable does not change'

@cody could you please assign this to someone having a deep knowledge of the loops and material system that could suggest the right way to make this work

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1294,22 +1295,29 @@ class FEProblemBase : public SubProblem, public Restartable
    * These methods are used to determine whether stateful material properties need to be stored on
-   * internal sides.  There are four situations where this may be the case: 1) DGKernels
-   * 2) IntegratedBCs 3)InternalSideUserObjects 4)ElementalAuxBCs
+   * internal sides.  There are five situations where this may be the case: 1) DGKernels
```
I would argue that there are 6; one could have InterfaceKernels without InterfaceUserObjects just as one can have `DGKernels` without `InternalSideUserObjects`

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1294,22 +1295,29 @@ class FEProblemBase : public SubProblem, public Restartable
    * @return Boolean indicating whether material properties need to be stored
    */
   bool needBoundaryMaterialOnSide(BoundaryID bnd_id, THREAD_ID tid);
+  bool needBoundaryMaterialOnInterafce(BoundaryID bnd_id, THREAD_ID tid);
```
misspelled

lindsayad commented on framework/src/loops/ComputeMaterialsObjectThread.C:
```diff
@@ -153,6 +153,96 @@ ComputeMaterialsObjectThread::onBoundary(const Elem * elem, unsigned int side, B
+    {
+      // Face Materials
+      if (_discrete_materials[Moose::FACE_MATERIAL_DATA].hasActiveBlockObjects(_subdomain, _tid))
+      {
```
Executing a single statement; you don't need the `{`. Same for other places in this method

lindsayad commented on framework/src/loops/ComputeMaterialsObjectThread.C:
```diff
@@ -153,6 +153,96 @@ ComputeMaterialsObjectThread::onBoundary(const Elem * elem, unsigned int side, B
+            face_n_points,
+            *neighbor,
+            neighbor_side,
+            true); // requried to spicfy wi are in the neighbor
```
misspellings

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -191,6 +198,51 @@ ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
+  const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();
+
+  if (!_interface_user_objects.hasActiveBoundaryObjects(bnd_id, _tid))
+  {
```
Don't need `{}`

arovinelli replied:
I fixed all of these kind of styles mistake

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -191,6 +198,51 @@ ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
+  }
+
+  _fe_problem.prepareFace(elem, _tid);
+  // TODO next line does not produce nay effect
```
Is this your comment? If it doesn't do anything, remove the commented lines. It may not have any effect because `reinitElemFace` has already been called...looks like this happens in `onBoundary`

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -191,6 +198,51 @@ ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
+  SwapBackSentinel face_sentinel(_fe_problem, &FEProblem::swapBackMaterialsFace, _tid);
+  _fe_problem.reinitMaterialsFace(elem->subdomain_id(), _tid);
+
+  // TODO activating the next two lines generate a segmentation fault, need to understand why
```
In `onBoundary` there is only one sentinel: the lines of code are:
```
  SwapBackSentinel sentinel(_fe_problem, &FEProblem::swapBackMaterialsFace, _tid);
  _fe_problem.reinitMaterialsFace(_subdomain, _tid);
  _fe_problem.reinitMaterialsBoundary(bnd_id, _tid);
```

lindsayad replied:
Perhaps that would resolve the segmentation fault?

arovinelli replied:
I tried to use was you suggested but it still result in a segfault. 
`_fe_problem.reinitMaterialsBoundary(bnd_id, _tid);` always cause a segfault. I wondering if we really need it, as the interface UO should not update the value of the boundary material property but only hold the quantities that will be used in computeQP

arovinelli replied:
@lindsayad  I went to the bottom of this and found what's happening:
The `reinitMaterialsBoundary` call for update stateful material boundary property, however the UO itself must be executed to update the stateful interface MP... therefore there is no value that can be pulled from the UO. The idea here is that stateful Interface MP are function of stateful and non stateful boundary and bulk properties, and should always be updated at the end. I think we might need to prevent the update of stateful interface MP.

lindsayad commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -246,19 +247,39 @@ MaterialPropertyStorage::initStatefulProps(MaterialData & material_data,
+    }
+    else
+    {
+      // if we are in teh neighbor we need to check that the property has not a null pointer
```
misspellings

lindsayad commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -385,8 +406,12 @@ void
 {
+  // TODO _stateful_prop_id_to_prop_id contains the stateful properties belonging to the master
+  // blockRestricted a more smart way to solve the issue we have is to create a proper
+  // _stateful_prop_id_to_prop_id for the nieghbor block
```
misspell

lindsayad replied:
Is there a large impediment to doing this? I don't want to introduce a lot of `TODO` statements in the code, but if there's a large overhead than this ok for now I would say

arovinelli replied:
I tried this route, but I'm I wasn't able to find a clean easy way to implement _stateful_prop_id_to_prop_id for the neighbour block. maybe you have some better idea.

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5350,6 +5385,20 @@ FEProblemBase::needBoundaryMaterialOnSide(BoundaryID bnd_id, THREAD_ID tid)
+  {
+    _bnd_mat_interface_cache[tid][bnd_id] = false;
+
+    if (_interface_user_objects.hasActiveBoundaryObjects(bnd_id, tid))
```
I want to see another call similar to what's in `FEProblemBase::needBoundaryMaterialOnSide`, something like:
```
if (_nl->needBoundaryMaterialOnInterface(bnd_id, tid))
```
and the above would check whether there are any active `_interface_kernel` objects

lindsayad commented on test/src/userobjects/InterfaceUO_QP.C:
```diff
@@ -0,0 +1,151 @@
+    mooseError("InterfaceUO_QP:: cannot fine the required element and side");
+}
+
+// void
```
Remove these commented lines

lindsayad reviewed (changes requested):
@arovinelli This is a very worthwhile capability that you're adding, and some of the internal changes are things that I will make use of. Thanks for taking the dive and exploring the framework.

lindsayad commented:
@arovinelli For your test that is failing pthreads and intel, specify `recover = false` in the `tests` file

lindsayad commented:
Have you generated a backtrace?

On Tue, Jun 26, 2018 at 12:52 PM, Andrea Rovinelli <notifications@github.com
> wrote:

> *@arovinelli* commented on this pull request.
> ------------------------------
>
> In framework/src/loops/ComputeUserObjectsThread.C
> <https://github.com/idaholab/moose/pull/11660#discussion_r198259815>:
>
> > +        (neighbor->level() < elem->level())))
> +  {
> +    return;
> +  }
> +
> +  _fe_problem.prepareFace(elem, _tid);
> +  // TODO next line does not produce nay effect
> +  // _fe_problem.reinitElemFace(elem, side, bnd_id, _tid);
> +  _fe_problem.reinitNeighbor(elem, side, _tid);
> +
> +  // Set up Sentinels so that, even if one of the reinitMaterialsXXX() calls throws, we
> +  // still remember to swap back during stack unwinding.
> +  SwapBackSentinel face_sentinel(_fe_problem, &FEProblem::swapBackMaterialsFace, _tid);
> +  _fe_problem.reinitMaterialsFace(elem->subdomain_id(), _tid);
> +
> +  // TODO activating the next two lines generate a segmentation fault, need to understand why
>
> I tried to use was you suggested but it still result in a segfault.
> _fe_problem.reinitMaterialsBoundary(bnd_id, _tid); always cause a
> segfault. I wondering if we really need it, as the interface UO should not
> update the value of the boundary material property but only hold the
> quantities that will be used in computeQP
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/11660#discussion_r198259815>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJxgcPbP20p1gmXDdvF-kX6aLOOcAVMGks5uAoMKgaJpZM4Ug-Lx>
> .
>

arovinelli commented:
yes. This what happens uncommenting lines 231 and 232
`#0  0x00007fffe9ed8277 in raise () from /lib64/libc.so.6
#1  0x00007fffe9ed9968 in abort () from /lib64/libc.so.6
#2  0x00007fffebd1a363 in __gnu_debug::_Error_formatter::_M_error (this=0x6294a0 <__gnu_debug::_Error_formatter::_M_at(char const*, unsigned int)::__formatter>)
    at ../../../../../gcc-7.2.0/libstdc++-v3/src/c++11/debug.cc:1069
#3  0x00007ffff784532f in std::__debug::vector<std::__debug::vector<double, std::allocator<double> >, std::allocator<std::__debug::vector<double, std::allocator<double> > > >::operator[] (this=0x99f1e8, __n=0) at /opt/moose/gcc-7.2.0/include/c++/7.2.0/debug/vector:424
#4  0x00007ffff782abfd in InterfaceUO_QP::getNewBoundaryPropertyValue (this=0x93b280, elem=0, side=2, qp=0)
    at /home/arovinelli/projects/moose/test/src/userobjects/InterfaceUO_QP.C:138
#5  0x00007ffff76f48bb in InterfaceUOMaterialStateful::computeQpProperties (this=0x98e4a0) at /home/arovinelli/projects/moose/test/src/materials/InterfaceUOMaterialStateful.C:45
#6  0x00007ffff5889c69 in Material::computeProperties (this=0x98e4a0) at /home/arovinelli/projects/moose/framework/src/materials/Material.C:245
#7  0x00007ffff588a167 in MaterialData::reinit (this=0x8f37b0, mats=...) at /home/arovinelli/projects/moose/framework/src/materials/MaterialData.C:70
#8  0x00007ffff59afcb0 in FEProblemBase::reinitMaterialsBoundary (this=0x8863a0, boundary_id=4, tid=0, swap_stateful=true)
    at /home/arovinelli/projects/moose/framework/src/problems/FEProblemBase.C:2453
#9  0x00007ffff5ff043f in ComputeUserObjectsThread::onInterface (this=0x7fffffffbd30, elem=0x768380, side=2, bnd_id=4)
    at /home/arovinelli/projects/moose/framework/src/loops/ComputeUserObjectsThread.C:232
#10 0x00007ffff5364b49 in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7fffffffbd30, 
    range=..., bypass_threading=false) at /home/arovinelli/projects/moose/framework/build/header_symlinks/ThreadedElementLoopBase.h:225
#11 0x00007ffff59e7855 in libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeUserObjectsThread> (
    range=..., body=...) at /home/arovinelli/projects/moose/scripts/../libmesh/installed/include/libmesh/threads_pthread.h:380
#12 0x00007ffff59b28f0 in FEProblemBase::computeUserObjects (this=0x8863a0, type=..., group=@0x7fffffffc48c: Moose::POST_AUX)
    at /home/arovinelli/projects/moose/framework/src/problems/FEProblemBase.C:2963
#13 0x00007ffff59a291a in FEProblemBase::initialSetup (this=0x8863a0) at /home/arovinelli/projects/moose/framework/src/problems/FEProblemBase.C:732
#14 0x00007ffff5ccd730 in Steady::init (this=0x922140) at /home/arovinelli/projects/moose/framework/src/executioners/Steady.C:50
#15 0x00007ffff6087df8 in MooseApp::executeExecutioner (this=0x6f7060) at /home/arovinelli/projects/moose/framework/src/base/MooseApp.C:799
#16 0x00007ffff60881a1 in MooseApp::run (this=0x6f7060) at /home/arovinelli/projects/moose/framework/src/base/MooseApp.C:904
#17 0x000000000041a808 in main (argc=3, argv=0x7fffffffc7b8) at /home/arovinelli/projects/moose/test/src/main.C:31
`

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -2713,6 +2713,12 @@ NonlinearSystemBase::needBoundaryMaterialOnSide(BoundaryID bnd_id, THREAD_ID tid
+bool
+NonlinearSystemBase::needBoundaryMaterialOnInterface(BoundaryID bnd_id, THREAD_ID tid) const
+{
+  return _integrated_bcs.hasActiveBoundaryObjects(bnd_id, tid);
```
This does not get to the root of the problem signified by the BISON segmentation fault (shown below) but I was envisioning this as `_interface_kernels.hasActiveBoundaryObjects`
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x000000010410e961 libmoose-dbg.0.dylib`MaterialPropertyStorage::initProps(this=0x000000011a8aeef0, material_data=0x000000011a8e2f88, elem=0x000000011a88bf60, side=0, n_qpoints=1, isNeighborOnInterface=true) at MaterialPropertyStorage.C:448
    frame #1: 0x000000010411064d libmoose-dbg.0.dylib`MaterialPropertyStorage::initStatefulProps(this=0x000000011a8aeef0, material_data=0x000000011a8e2f88, mats=size=1, n_qpoints=1, elem=0x000000011a88bf60, side=0, isNeighborOnInterface=true) at MaterialPropertyStorage.C:230
    frame #2: 0x000000010405809f libmoose-dbg.0.dylib`ComputeMaterialsObjectThread::onInterface(this=0x00007fff5fbfe710, elem=0x000000011a88bed0, side=1, bnd_id=16) at ComputeMaterialsObjectThread.C:224
    frame #3: 0x000000010424fb1d libmoose-dbg.0.dylib`ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator(this=0x00007fff5fbfe710, range=0x00000001171362a0, bypass_threading=true)(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) at ThreadedElementLoopBase.h:225
    frame #4: 0x00000001047e5f53 libmoose-dbg.0.dylib`FEProblemBase::initialSetup(this=0x000000011a003a18) at FEProblemBase.C:587
    frame #5: 0x0000000103dd5657 libmoose-dbg.0.dylib`Transient::init(this=0x000000011a906208) at Transient.C:295
    frame #6: 0x0000000105012e03 libmoose-dbg.0.dylib`MooseApp::executeExecutioner(this=0x0000000118809a18) at MooseApp.C:799
    frame #7: 0x000000010501362b libmoose-dbg.0.dylib`MooseApp::run(this=0x0000000118809a18) at MooseApp.C:904
    frame #8: 0x000000010000221c bison-dbg`main(argc=3, argv=0x00007fff5fbff818) at main.C:49
    frame #9: 0x00007fffcd80c235 libdyld.dylib`start + 1
    frame #10: 0x00007fffcd80c235 libdyld.dylib`start + 1
```

lindsayad replied:
Making this change will consequently make the BISON test pass...

arovinelli replied:
@lindsayad 
your suggested change makes total sense. However after applying it and running tests on my machine the same error generated in BISON appears in MOOSE for
`interfacekernels/1d_interface.test ............................................................ FAILED (CRASH)`
with BT
```
#0  0x00007ffff588e91f in MaterialPropertyStorage::initProps (this=0x879b40, material_data=..., elem=..., 
    side=0, n_qpoints=1, isNeighborOnInterface=true)
    at /home/arovinelli/projects/moose/framework/src/materials/MaterialPropertyStorage.C:463
#1  0x00007ffff588d3cf in MaterialPropertyStorage::initStatefulProps (this=0x879b40, material_data=..., 
    mats=..., n_qpoints=1, elem=..., side=0, isNeighborOnInterface=true)
    at /home/arovinelli/projects/moose/framework/src/materials/MaterialPropertyStorage.C:230
#2  0x00007ffff5fea720 in ComputeMaterialsObjectThread::onInterface (this=0x7fffffffc120, elem=0x86ad20, 
    side=1, bnd_id=2) at /home/arovinelli/projects/moose/framework/src/loops/ComputeMaterialsObjectThread.C:224
#3  0x00007ffff5364021 in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7fffffffc120, range=..., bypass_threading=true)
    at /home/arovinelli/projects/moose/framework/build/header_symlinks/ThreadedElementLoopBase.h:225
#4  0x00007ffff59a2deb in FEProblemBase::initialSetup (this=0x897610)
    at /home/arovinelli/projects/moose/framework/src/problems/FEProblemBase.C:587
#5  0x00007ffff5cce092 in Steady::init (this=0x92f4c0)
    at /home/arovinelli/projects/moose/framework/src/executioners/Steady.C:50
#6  0x00007ffff6087c56 in MooseApp::executeExecutioner (this=0x6f7080)
    at /home/arovinelli/projects/moose/framework/src/base/MooseApp.C:799
#7  0x00007ffff6087fff in MooseApp::run (this=0x6f7080)
    at /home/arovinelli/projects/moose/framework/src/base/MooseApp.C:904
#8  0x000000000041a808 in main (argc=3, argv=0x7fffffffc7c8)
    at /home/arovinelli/projects/moose/test/src/main.C:31
```
this means that the safeguard I implemented is not working and that `_stateful_prop_id_to_prop_id` in `MaterialPropertyStorage` is not behaving as I was thinking.
There is something else that need to be addressed here. Do you have any clue about the context of `_stateful_prop_id_to_prop_id` (is it defined by element, boundary, block,...)?

lindsayad replied:
Have you looked in a debugger at this? Your check is working, but `material_data.props()[0]` is a `nullptr` so calling a method through it results in `EXEC_BAD_ACCESS`. As to the underlying cause, I'm going to investigate a bit

arovinelli commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -400,15 +425,36 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
+    else // if so we need check wich material property exist in the current element
+    {
+      if (props(&elem, side)[i] == nullptr && prop_id < props(&elem, side).size() &&
+          material_data.props()[prop_id] != nullptr)
```
adding `material_data.props()[prop_id] != nullptr` makes all test work

arovinelli commented:
@lindsayadAlex, thanks a lot for the hint! I think I've fixed it!
Let's wait for this run of tests to finish but BISON already past them
all (finger crossed).

Andrea

On Thu, 2018-06-28 at 07:46 -0700, Alex Lindsay wrote:
> @lindsayad commented on this pull request.
> In framework/src/systems/NonlinearSystemBase.C:
> > @@ -2713,6 +2713,12 @@
> NonlinearSystemBase::needBoundaryMaterialOnSide(BoundaryID bnd_id,
> THREAD_ID tid
>    return _integrated_bcs.hasActiveBoundaryObjects(bnd_id, tid);
>  }
>  
> +bool
> +NonlinearSystemBase::needBoundaryMaterialOnInterface(BoundaryID
> bnd_id, THREAD_ID tid) const
> +{
> +  return _integrated_bcs.hasActiveBoundaryObjects(bnd_id, tid);
> Have you looked in a debugger at this? Your check is working, but
> material_data.props()[0] is a nullptr so calling a method through it
> results in EXEC_BAD_ACCESS. As to the underlying cause, I'm going to
> investigate a bit
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

arovinelli commented:
@lindsayad and @permcody 
the two test failing are restart for `coupled_value_coupled_flux.i`, should i simply prevent them for checking restart as I did for my test?

arovinelli commented:
@permcody your right and I agree on your plan. 
I have no experience with the recover system and output format, could you point me where to start looking (e.g. which are the basic functions and methods used for dumping recover data)?

lindsayad commented:
The real issue is a bit insidious. Here are my notes on the problem. MOOSE has `_material_data`, `_bnd_material_data`, and `_neighbor_material_data`. Let's summarize what the different materials in that input file do to the `MaterialData` objects:

- `StatefulMaterial` adds diffusivity to `_bnd_material_data`
- The "general" `GenericConstantMaterial` adds its single property to `_material_data` (for volume), then to `_bnd_material_data` (for face), then to `_neighbor_material_data` (for neighbor)
- The "boundary" `GenericConstantMaterial` add its two properties to `_bnd_material_data`

Now, there are two `MaterialPropertyStorage` objects: `_material_props` and `bnd_material_props`. `_material_data` is initialized with `_material_props`
while the other two material data objects are intialized with
`_bnd_material_props`. **Note** that the object names I'm using here are their names in `FEProblemBase` which is where the objects are created.

Now let's summarize the state of the `MaterialData` and `MaterialPropertyStorage` objects after declaring all our material properties:
```
static MaterialPropertyStorage member _props_ids:
"diffusivity", 0
"dummy", 1
"D", 2
"D_neighbor", 3

_bnd_material_props (MaterialPropertyStorage): _stateful_prop_id_to_prop_id
[0] : 0

_material_props (MaterialPropertyStorage): _stateful_prop_id_to_prop_id
Empty

_material_data (uses _material_props)
[0] : "diffusivity" : null
[1] : "dummy" : non-null

_bnd_material_data (uses _bnd_material_props)
[0] : "diffusivity" : non-null
[1] : "dummy" : non-null
[2] : "D" : non-null
[3] : "D_neighbor" : non-null

_neighbor_material_data (uses _bnd_material_props)
[0] : "diffusivity" : null
[1] : "dummy" : non-null
```
The root problem is this: creating a stateful property in a boundary `Material` creates a stateful property for `_neighbor_material_data` as well as for `_bnd_material_data` because they both use the same underlying `MaterialPropertyStorage` object :-(

lindsayad commented:
@andrsd @permcody Thoughts on a `_neighbor_material_props` `MaterialPropertyStorage` instance?

arovinelli commented:
@lindsayad thanks a lot for figuring out the root cause or the problem, this is really helpful.
I might be dumb :(, but even with your help I'm not able to find the exact code line where the stateful boundary property is added to the `_neighbor_material_data` structure. Could you please point that out to me?

lindsayad commented:
The stateful boundary property *ID* is not added to `_neighbor_material_data` per se. It is added to the `MaterialPropertyStorage` member `_storage`. The relevant stack trace (on MOOSE devel), intitiated by a call from `getMaterialPropertyOld` in any standard derived `Material` is:
```
  * frame #0: 0x00000001012cd6c0 libmoose-dbg.0.dylib`MaterialPropertyStorage::addPropertyOld(this=0x000000010e13adf0, prop_name="diffusivity") at MaterialPropertyStorage.C:350
    frame #1: 0x0000000100181b6e libmoose_test-dbg.0.dylib`MaterialProperty<double>& MaterialData::getPropertyOld<double>(this=0x000000010e17e078, name="diffusivity") at MaterialData.h:256
    frame #2: 0x00000001001813d3 libmoose_test-dbg.0.dylib`MaterialProperty<double> const& MaterialPropertyInterface::getMaterialPropertyOldByName<double>(this=0x000000010e80b8e8, name=0x00007fff5fbfafd0) at MaterialPropertyInterface.h:378
    frame #3: 0x000000010035c6cf libmoose_test-dbg.0.dylib`MaterialProperty<double> const& Material::getMaterialPropertyOldByName<double>(this=0x000000010e80b418, prop_name="diffusivity") at Material.h:350
    frame #4: 0x0000000100318b62 libmoose_test-dbg.0.dylib`MaterialProperty<double> const& Material::getMaterialPropertyOld<double>(this=0x000000010e80b418, name="diffusivity") at Material.h:315
```

arovinelli commented:
@permcody and @lindsayad   I was thinking to your comment about the failing recover test...
There is only one step in the exodus because it fails to recover so I guess it just read the mesh save it and then execute... (see below)
Now, why is failing to recover it's possibly due to the fact that the fix I have implemented generated a data structure for stateful material boundary property that is not compatible with restart (there are hanging nullptrs)

The call is triggered by 

```
bool
FEProblemBase::needBoundaryMaterialOnInterface(BoundaryID bnd_id, THREAD_ID tid)
{
  if (_bnd_mat_interface_cache[tid].find(bnd_id) == _bnd_mat_interface_cache[tid].end())
  {
    _bnd_mat_interface_cache[tid][bnd_id] = false;

```
    **if (_nl->needBoundaryMaterialOnInterface(bnd_id, tid))
    _bnd_mat_interface_cache[tid][bnd_id] = true;**
``` 
   else
    if (_interface_user_objects.hasActiveBoundaryObjects(bnd_id, tid))
      _bnd_mat_interface_cache[tid][bnd_id] = true;
  }

  return _bnd_mat_interface_cache[tid][bnd_id];
}
```
Note that I tried and commenting the starred lines the test will pass.

I guess this issue will be resolved by fixing #11766,
so if you are ok knowing why this is happening you can resume review.


```
interfacekernels/1d_interface.test: Working Directory: /opt/civet/build_0/moose/test/tests/interfacekernels/1d_interface
interfacekernels/1d_interface.test: Running command: /opt/civet/build_0/moose/test/moose_test-opt -i coupled_value_coupled_flux.i --recover --recoversuffix cpr --error --error-unused --error-override --no-gdb-backtrace
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: Using coupled_value_coupled_flux_out_cp/0000 for recovery.
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: Cannot recover steady solves!
interfacekernels/1d_interface.test: Exiting...
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: Running exodiff: /opt/civet/build_0/moose/framework/contrib/exodiff/exodiff -m  -F 1e-10 -t 5.5e-06  /opt/civet/build_0/moose/test/tests/interfacekernels/1d_interface/gold/coupled_value_coupled_flux_out.e /opt/civet/build_0/moose/test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_out.e
interfacekernels/1d_interface.test: ERROR: 
interfacekernels/1d_interface.test:    *****************************************************************
interfacekernels/1d_interface.test:               EXODIFF	(Version: 2.90) Modified: 2018-02-15
interfacekernels/1d_interface.test:               Authors:  Richard Drake, rrdrake@sandia.gov           
interfacekernels/1d_interface.test:                         Greg Sjaardema, gdsjaar@sandia.gov          
interfacekernels/1d_interface.test:               Run on    2018/06/28   14:57:31 MDT
interfacekernels/1d_interface.test:    *****************************************************************
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: Reading first file ... 
interfacekernels/1d_interface.test: Reading second file ... 
interfacekernels/1d_interface.test:   FILE 1: /opt/civet/build_0/moose/test/tests/interfacekernels/1d_interface/gold/coupled_value_coupled_flux_out.e
interfacekernels/1d_interface.test:    Title: my_coupled_tied_value_out.e
interfacekernels/1d_interface.test:           Dim = 3, Blocks = 2, Nodes = 11, Elements = 10, Nodesets = 4, Sidesets = 4
interfacekernels/1d_interface.test:           Vars: Global = 0, Nodal = 2, Element = 0, Nodeset = 0, Sideset = 0, Times = 2
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test:   FILE 2: /opt/civet/build_0/moose/test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_out.e
interfacekernels/1d_interface.test:    Title: coupled_value_coupled_flux_out.e
interfacekernels/1d_interface.test:           Dim = 3, Blocks = 2, Nodes = 11, Elements = 10, Nodesets = 4, Sidesets = 4
interfacekernels/1d_interface.test:           Vars: Global = 0, Nodal = 2, Element = 0, Nodeset = 0, Sideset = 0, Times = 1
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: exodiff: ERROR: .. First file has 2 result times while the second file has 1.
interfacekernels/1d_interface.test: exodiff: INFO .. Map option is not needed.
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: Nodal coordinates will be compared .. tol:    1e-06 (absolute), floor:        0
interfacekernels/1d_interface.test: Time step values will be compared .. tol:  5.5e-06 (relative), floor:    1e-10
interfacekernels/1d_interface.test: No Global variables on either file.
interfacekernels/1d_interface.test: Nodal variables to be compared:
interfacekernels/1d_interface.test: 	u                                tol:  5.5e-06 (relative), floor:    1e-10
interfacekernels/1d_interface.test: 	v                                      5.5e-06 (relative),           1e-10
interfacekernels/1d_interface.test: No Element variables on either file.
interfacekernels/1d_interface.test: No Element Attribute variables on either file.
interfacekernels/1d_interface.test: No Nodeset variables on either file.
interfacekernels/1d_interface.test: No Sideset variables on either file.
interfacekernels/1d_interface.test: Sideset Distribution Factors will be compared .. tol:    1e-06 (relative), floor:        0
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test:   ==============================================================
interfacekernels/1d_interface.test:   NOTE: All node and element ids are reported as global ids.
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: Sideset Distribution Factors:
interfacekernels/1d_interface.test:   --------- Time step 1, 0.0000000e+00 ~ 0.0000000e+00, rel diff:  0.00000e+00 ---------
interfacekernels/1d_interface.test: Nodal variables:
interfacekernels/1d_interface.test: 
interfacekernels/1d_interface.test: exodiff: Files are different (# time steps differ)
interfacekernels/1d_interface.test:  
interfacekernels/1d_interface.test: ################################################################################
interfacekernels/1d_interface.test: Tester failed, reason: EXODIFF
interfacekernels/1d_interface.test: 
```

arovinelli commented:
@lindsayad thanks for taking care of  #11766 
@permcody after merging  #11778 recover works (at least on my machine) so everything should be fine now

lindsayad reviewed (changes requested):
@arovinelli Thanks for being patient during the review process. This is great work. Most of my requested changes are spelling and style. Might be helpful to have a discussion with @permcody and @andrsd about the user object and material dependency resolution.

arovinelli commented on framework/src/materials/MaterialData.C:
```diff
@@ -64,10 +64,16 @@ MaterialData::swap(const Elem & elem, unsigned int side /* = 0*/)
   for (const auto & mat : mats)
-    mat->computeProperties();
+  {
+    if (prevent_update_interface_materials && mat->isInterfaceMaterial())
```
@lindsayad thanks for reviewing this PR. A clean way to avoid these checks would be to distinguish between boundary and interface materials,e.g. adding  `_interface_material_data`.
Then one can write a new `reinitMaterialsInterface`. I'm not sure about the overhead of this.

lindsayad replied:
I do think that would be a solution, and I think it's reasonable. However, I'm not sure @andrsd would like it. The current design is that boundary and face materials both use the same `MaterialData` object (see below). If one was to add `_interface_material_data`, I think you could argue that we should also split `_bnd_material_data` out into `_bnd_material_data` and `_internal_face_material_data`, and I'm just not sure if that's what we want. I believe @andrsd gets back next Monday, and I hope we can resolve it then. Hopefully you can wait until then?
```
std::shared_ptr<MaterialData>
FEProblemBase::getMaterialData(Moose::MaterialDataType type, THREAD_ID tid)
{
  std::shared_ptr<MaterialData> output;
  switch (type)
  {
    case Moose::BLOCK_MATERIAL_DATA:
      output = _material_data[tid];
      break;
    case Moose::NEIGHBOR_MATERIAL_DATA:
      output = _neighbor_material_data[tid];
      break;
    case Moose::BOUNDARY_MATERIAL_DATA:
    case Moose::FACE_MATERIAL_DATA:
      output = _bnd_material_data[tid];
      break;
  }
  return output;
}
```

arovinelli replied:
@lindsayad yes I can wait no problem. Now everything is working fine and I can  work on a branch of this PR for doing other stuff. As far as I'm concerned now it's just a matter of making the code cleaner. Also, I want  this to be done as correct as possible as I would like to avoid debugging dependency problems in the future :). It has been a long journey since the beginning but now I can see the light at the end of tunnel and start doing interesting calculations with MOOSE :)

arovinelli commented:
@andrsd did you get a chance to look into this PR?

lindsayad commented:
@andrsd and I talked about this extensively yesterday. We decided that you should go back to your initial path...we think we should make a `MaterialBase` class that the existing `Material` and a new `InterfaceMaterial` inherits from. The `InterfaceMaterial` will inherit from `NeighborCoupleable` and `TwoMaterialPropertyInterface` allowing it to pull both element and neighbor variable values and material properties. Then you do not have to pull from a `UserObject` and we do not have to resolve dependency issues that we are currently incapable of resolving. You are welcome to work on this if you would like. Otherwise I will get to it at some point.

`InterfaceUserObject` is still something of value that we want in the framework. So we still want this PR. But the `Material` aspects of this PR should be pulled out.

Sorry that we lead you away from your initial path and are now suggesting you return to it!

lindsayad commented:
We keep the dependencies within the material objects instead of splitting out to user objects. onInterface will be called after onElement, etc just as it is done in all our loop objects. We may need an _interface_material_data, we will see. But this design will avoid the need to have a material input parameter like interface_material = true. 

> On Jul 13, 2018, at 8:28 AM, Andrea Rovinelli <notifications@github.com> wrote:
> 
> Dear @lindsayad, thanks for taking the time to discuss this with @andrsd .
> I'm not sure I understand your view. @andrsd suggested to switch to a UO to avoid pulling incorrect data (i.e. the master material beng current while the neighbor not).
> 
> To do what you and David suggested is not just subclassing a 'MaterialBase' and create an InterfaceMaterial, we still need a ComputeMaterialsOnInterface to figure out the right dependence resolution (as the interface property still need to be computed later in the sequence). We also need to add an _interface_material_data structure and INTERFACE_MATERIAL_DATA type.
> 
> Which is basically what we discussed a few post above.
> Is this the action plan you and david come out with, or I'm missing something?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

arovinelli commented:
I understand what you are saying. I might dedicate some time to it, however deadline are coming up and I can't work on this full time. If I'll come out with something reasonable, I'll might make a new request for it (actually we might open it even now such that it is in the todo list and we can discuss further details).
In the meanwhile I'll take some time to clear this PR from the material dependence and come out with some new test, such that this can be merged.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/212389/) on 30d6240 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11660/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11660/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aa6cfa458ad0dd8a9a12bf06f88fecd2ef3422d6`

arovinelli commented:
@lindsayad and @andrsd I was able to make this UO work taking out all the strange material dependencies (`is_interrface_material` does not exist anymore). I simply updated boundary materials after the interface uo excecution. All tests works.
Please see if you like this.

lindsayad commented on framework/src/loops/ComputeUserObjectsThread.C:
```diff
@@ -191,6 +198,40 @@ ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
+  for (const auto & uo : objects)
+    uo->execute();
+
+  _fe_problem.reinitMaterialsBoundary(bnd_id, _tid);
```
This is unfortunately something that works for your particular use case, e.g. compute face and neighbor material properties, operate on them in the user object, and then pass them into a boundary restricted material. But what if someone wanted to create an `InterfaceUserObject` that has boundary restricted material properties as a pre-requisite instead of the reverse? Our precedent is to evaluate material properties first and then evaluate user objects, and I think we should stick to it. Hence our belief that we should add `InterfaceMaterial`

YaqiWang commented:
Can this be get into moose without `InterfaceMaterial`?

lindsayad commented:
Yes, but all the `Material` specific pieces of this PR should be removed

On Thu, Jul 19, 2018 at 12:44 PM, Yaqi <notifications@github.com> wrote:

> Can this be get into moose without InterfaceMaterial?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/11660#issuecomment-406376112>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJxgcEzVHxTsnmM3ny4Kdwb1Drf4pQ_Jks5uINNzgaJpZM4Ug-Lx>
> .
>

arovinelli commented:
If you do that this UO will not be able anymore to be used on interface to grep neighbor MP.
We could leave it as it is but making sure that this object only works for grabbing volumetric MP and delivering to Boundary materials, Otherwise it would be nothing more than a post processor.

lindsayad commented:
There are a lot of user objects that are far more than postprocessors that don't rely on reinitializing materials after the user objects are computed. You are thinking only about your own use case!!!!!

I don't know what limitation you're talking about with an `InterfaceMaterial`. In `ComputeMaterialsObjectThread` or any loop that inherits from `ThreadedElementLoopBase` we call in this order:

1. `onElement`
2. `onBoundary`
3. `onInternalSide`
4. `onInterface`

So if you had an `InterfaceKernel` and any `Kernels` at all, `ComputeResidualThread` would reinit the volume materials in `onElement`, and then the interface material properties would be reinitialized in `onInterface`, and you would know that all the volume properties are prepared and ready for use by the interface material properties. Moreover, if you wanted to be sure that your volume properties are reinitialized regardless of whether you have any `Kernels` in your system, we could add an `_fe_problem.reinitMaterials` before `_fe_problem.reinitMaterialsInterface` in `ComputeResidualThread::onInterface`

arovinelli commented:
So I started working on this again. I'm trying to take out the material dependence as request by @lindsayad . If completely  comment `ComputeMaterialsObjectThread::onInterface` this function the UO correctly compute variables accross the interface but does not get proper value for material properties. 
Any suggestion?

arovinelli commented:
@permcody @lindsayad @jiangwen84, can anyone reopen this PR please?

arovinelli commented:
Thanks Cody!
I'll do that and avoid force push in the future!

arovinelli commented:
moved to new PR #12079
